### PR TITLE
python3Packages.ibis-framework: 3.2.0 -> 4.0.0

### PR DESCRIPTION
**Pull request by lilyinstarlight**
###### Description of changes

While investigating why there was a new failure in #211155, I basically figured out why this package in particular was failing to build (despite it turning out to have nothing to do with that PR). So I had some spare time and updated some of the deps to get it to build and seemingly function (including fixing some of the Python 3.11 fallout)

I don't actually use ibis-framework though so @cpcloud or someone should definitely test all of this

Note:
This requires #211015 to build `python3{10,11}Packages.ibis-framework`, so I've targeted this towards staging, where that PR is (but I can change it if desired -- just note that `ibis-framework` won't build until the next staging-next anyway if this does go to master)

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by veprbl — COMMENTED** [review]
`pyarrow` change is fine by me

File: pkgs/development/python-modules/ibis-framework/default.nix [hunk 1]
```diff
@@ -93,31 +98,27 @@ buildPythonPackage rec {
 
   checkInputs = [
     pytestCheckHook
+    black
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/ibis-framework/default.nix:101]: Please patch out the black tests. Updating black should not trigger to many rebuilds or cause regressions.

Comment (lilyinstarlight) [reply pkgs/development/python-modules/ibis-framework/default.nix:101]: They have a whole util method for that now, apparently, which compares decompiled code via pytest-snapshot. So I adjusted the derivation in `postPatch` to do a sed removal for calls to that util method and then to fix up the remaining tests that manually specify `format=True` on `ibis.decompile` (defined in `ibis/tests/util.py` and what specifically uses black)

**Comment by lilyinstarlight** [comment]
That last push was a rebase to fix conflicts. The changes for fixing conflicts were renaming `checkInputs` to `nativeCheckInputs` and dropping the `python3Packages.parameterized` change since that got fixed separately by #211974

**Review by SuperSandro2000 — APPROVED** [review]

**Pull request merged**